subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -115,180 +109,81 @@ public ClassSetDiff findChangesSince(ClassSetAnalysis other) {
-                                         Set<GeneratedResource> resultResources,
-                                         Iterable<String> privateDependentClasses,
-                                         Iterable<String> accessibleDependentClasses) {
-
-        for (String privateDependentClass : privateDependentClasses) {
-            if (!visitedClasses.add(privateDependentClass)) {
+        while (!remaining.isEmpty()) {
+            String accessibleDependent = remaining.pop();
+            if (!visited.add(accessibleDependent)) {
                 continue;
             }
-            privateResultClasses.add(privateDependentClass);
-            DependentsSet currentDependents = getDependents(privateDependentClass);
-            if (!currentDependents.isDependencyToAll()) {
-                resultResources.addAll(currentDependents.getDependentResources());
-            }
-        }
-
-        processTransitiveDependentClasses(visitedClasses, accessibleResultClasses, resultResources, accessibleDependentClasses);
-    }
-
-    private void processTransitiveDependentClasses(Set<String> visitedClasses,
-                                                   Set<String> accessibleResultClasses,
-                                                   Set<GeneratedResource> resultResources,
-                                                   Iterable<String> accessibleDependentClasses) {
-        Deque<String> remainingAccessibleDependentClasses = new ArrayDeque<>();
-        for (String accessibleDependentClass : accessibleDependentClasses) {
-            remainingAccessibleDependentClasses.add(accessibleDependentClass);
-        }
-
-        while (!remainingAccessibleDependentClasses.isEmpty()) {
-            String accessibleDependentClass = remainingAccessibleDependentClasses.pop();
-            if (!visitedClasses.add(accessibleDependentClass)) {
-                continue;
+            accessibleDependents.add(accessibleDependent);
+            remaining.addAll(classDependenciesFromAnnotationProcessing.get(accessibleDependent));
+            dependentResources.addAll(resourceDependenciesFromAnnotationProcessing.get(accessibleDependent));
+            DependentsSet dependents = findDirectDependents(accessibleDependent);
+            if (dependents.isDependencyToAll()) {
+                return dependents;
             }
-            accessibleResultClasses.add(accessibleDependentClass);
-            DependentsSet currentDependents = getDependents(accessibleDependentClass);
-            if (!currentDependents.isDependencyToAll()) {
-                resultResources.addAll(currentDependents.getDependentResources());
-                remainingAccessibleDependentClasses.addAll(currentDependents.getAccessibleDependentClasses());
+            for (String dependentClass : dependents.getAllDependentClasses()) {
+                remaining.addAll(classDependenciesFromAnnotationProcessing.get(dependentClass));
```

Comment:
This was missing from the previous algorithm, which meant that recompile chains would often abruptly end at annotation processing boundaries. It now keeps going across these boundaries, assuming that any change that affects the annotated class could potentially result in a completely different API on the generated classes. I've added an integration test to keep this from regressing.